internal/cms/signed.go

```diff
@@ -0,0 +1,317 @@
+	Content []byte
+
+	// ContentType is the content type of the EncapsulatedContentInfo.
+	ContentType asn1.ObjectIdentifier
+
+	// Certificates is the list of certificates in the SignedData.
+	Certificates []*x509.Certificate
+
+	// CRLs is the list of certificate revocation lists in the SignedData.
+	CRLs []x509.RevocationList
+
+	// SignerInfos is the list of signer information in the SignedData.
+	SignerInfos []SignerInfo
+}
+
+// ParseSignedData parses ASN.1 BER-encoded SignedData structure to golang
+// friendly types.
+func ParseSignedData(berData []byte) (*ParsedSignedData, error) {
+	data, err := ber.ConvertToDER(berData)
+	if err != nil {
+		return nil, SyntaxError{Message: "invalid signed data: failed to convert from BER to DER", Detail: err}
+	}
+	var contentInfo ContentInfo
+	if _, err := asn1.Unmarshal(data, &contentInfo); err != nil {
+		return nil, SyntaxError{Message: "invalid content info: failed to unmarshal DER to ContentInfo", Detail: err}
+	}
+	if !oid.SignedData.Equal(contentInfo.ContentType) {
+		return nil, ErrNotSignedData
+	}
+
+	var signedData SignedData
+	if _, err := asn1.Unmarshal(contentInfo.Content.Bytes, &signedData); err != nil {
+		return nil, SyntaxError{Message: "invalid signed data", Detail: err}
+	}
+	certs, err := x509.ParseCertificates(signedData.Certificates.Bytes)
+	if err != nil {
+		return nil, SyntaxError{Message: "failed to parse X509 certificates from signed data", Detail: err}
+	}
+
+	return &ParsedSignedData{
+		Content:      signedData.EncapsulatedContentInfo.Content,
+		ContentType:  signedData.EncapsulatedContentInfo.ContentType,
+		Certificates: certs,
+		CRLs:         signedData.CRLs,
+		SignerInfos:  signedData.SignerInfos,
+	}, nil
+}
+
+// Verify attempts to verify the content in the parsed signed data against the signer
+// information. The `Intermediates` in the verify options will be ignored and
+// re-contrusted using the certificates in the parsed signed data.
+// If more than one signature is present, the successful validation of any signature
+// implies that the content in the parsed signed data is valid.
+// On successful verification, the list of signing certificates that successfully
+// verify is returned.
+// If all signatures fail to verify, the last error is returned.
+//
+// References:
+//   - RFC 5652 5   Signed-data Content Type
+//   - RFC 5652 5.4 Message Digest Calculation Process
+//   - RFC 5652 5.6 Signature Verification Process
+//
+// WARNING: this function doesn't do any revocation checking.
+func (d *ParsedSignedData) Verify(ctx context.Context, opts x509.VerifyOptions) ([][]*x509.Certificate, error) {
+	if len(d.SignerInfos) == 0 {
+		return nil, ErrSignerInfoNotFound
+	}
+	if len(d.Certificates) == 0 {
+		return nil, ErrCertificateNotFound
+	}
+
+	intermediates := x509.NewCertPool()
+	for _, cert := range d.Certificates {
+		intermediates.AddCert(cert)
+	}
+	opts.Intermediates = intermediates
+	verifiedSignerMap := map[string][]*x509.Certificate{}
```

Comment:
Question: What's the meaning of having a map instead of a slice?

Reply:
Good question. Because it is possible that multiple `SignerInfo` structures use the same signing certificate, using a map can help avoid duplicate certificates.